tests/e2e/README.md

```diff
@@ -24,11 +24,90 @@ core bootstrapping logic that creates a testing environment via Docker
 containers. A testing network is created dynamically with 2 test
 validators.
 
-The file e2e\_test.go contains the actual end-to-end integration tests
+The file `e2e_test.go` contains the actual end-to-end integration tests
 that utilize the testing suite.
 
-Currently, there is a single test in `e2e_test.go` to query the balances
-of a validator.
+Currently, there is a single IBC test in `e2e_test.go`.
+
+Additionally, there is an ability to disable certain components
+of the e2e suite. This can be done by setting the environment
+variables. See "Environment variables" section below for more details.
+
+## How It Works
+
+Conceptually, we can split the e2e setup into 2 parts:
+
+1. Chain Initialization
+
+    The chain can either be initailized off of the current branch, or off the prior mainnet release and then upgraded to the current branch.
+
+    If current, we run chain initialization off of the current Git branch
+    by calling `chain.Init(...)` method in the `configurer/current.go`.
+
+    If with the upgrade, the same `chain.Init(...)` function is run inside a Docker container
+    of the previous Osmosis version, inside `configurer/upgrade.go`. This is
+    needed to initialize chain configs and the genesis of the previous version that
+    we are upgrading from.
+
+    The decision of what configuration type to use is decided by the `Configurer`.
+    This is an interface that has `CurrentBranchConfigurer` and `UpgradeConfigurer` implementations.
+    There is also a `BaseConfigurer` which is shared by the concrete implementations. However,
+    The user of the `configurer` package does not need to know about this detail.
+
+    When the desired configurer is created, the caller may
+    configure the chain in the desired way as follows:
+
+    ```go
+    conf, _ := configurer.New(..., < isIBCEnabled bool >, < isUpgradeEnabled bool >)
+
+    conf.ConfigureChains()
+    ```
+
+    The caller (e2e setup logic), does not need to be concerned about what type of
+    configurations is hapenning in the background. The appropriate logic is selected
+    depending on what the values of the arguments to `configurer.New(...)` are.
+
+    The configurer constructor is using a factory design pattern
+    to decide on what kind of configurer to return.
```

Comment:
This pattern encapsulates the struct instantiation logic to avoid exposing it to the client:
https://www.tutorialspoint.com/design_pattern/factory_pattern.htm

I will add a comment about this